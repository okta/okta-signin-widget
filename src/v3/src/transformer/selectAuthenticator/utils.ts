/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { Input } from '@okta/okta-auth-js';
import { IdxOption } from '@okta/okta-auth-js/types/lib/idx/types/idx-js';

import {
  AUTHENTICATOR_ALLOWED_FOR_OPTIONS,
  AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP,
  AUTHENTICATOR_KEY,
} from '../../constants';
import { ActionParams, AuthenticatorButtonElement, ButtonType } from '../../types';
import { loc } from '../../util';

export const getOptionValue = (
  inputs: Input[],
  key: string,
): Input | undefined => inputs?.find(
  ({ name }) => name === key,
);

const getAuthenticatorDataSeVal = (authenticatorKey: string, methodType?: string): string => {
  if (authenticatorKey) {
    const method = methodType ? `-${methodType}` : '';
    return `${authenticatorKey}${method}`;
  }
  return '';
};

const reorderAuthenticatorButtons = (
  authButtons: AuthenticatorButtonElement[],
  deviceKnown?: boolean,
): AuthenticatorButtonElement[] => {
  if (authButtons.length <= 1) {
    return authButtons;
  }
  const fastpassAuthenticator = authButtons.find(
    (button) => button.options.actionParams?.['authenticator.methodType'] === 'signed_nonce',
  );
  if (!fastpassAuthenticator) {
    return authButtons;
  }

  const updatedAuthenticatorBtns = authButtons.filter(
    (button) => button.options.actionParams?.['authenticator.methodType'] !== 'signed_nonce',
  );

  // Re-arrange fastpass in options based on deviceKnown
  // If deviceKnown is set, set fastpass as the first option in the list
  // otherwise, place it as the last item in the list of OV options
  if (deviceKnown) {
    updatedAuthenticatorBtns.unshift(fastpassAuthenticator);
  } else {
    updatedAuthenticatorBtns.push(fastpassAuthenticator);
  }

  return updatedAuthenticatorBtns;
};

const buildOktaVerifyOptions = (
  options: IdxOption[],
  step: string,
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const ovRemediation = options.find((option) => option.relatesTo?.key === AUTHENTICATOR_KEY.OV);
  const id = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'id')?.value;
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType?.options?.length) {
    return [];
  }

  return methodType.options.map((option: IdxOption, index: number) => {
    const authenticatorButton: AuthenticatorButtonElement = {
      type: 'AuthenticatorButton',
      label: option.label,
      id: `auth_btn_${AUTHENTICATOR_KEY.OV}_${option.value || id}`,
      options: {
        type: ButtonType.BUTTON,
        key: AUTHENTICATOR_KEY.OV,
        ctaLabel: isEnroll
          ? loc('oie.enroll.authenticator.button.text', 'login')
          : loc('oie.verify.authenticator.button.text', 'login'),
        description: isEnroll
          ? loc(AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[AUTHENTICATOR_KEY.OV], 'login')
          : loc('oie.okta_verify.label', 'login'),
        actionParams: {
          'authenticator.methodType': option.value,
          'authenticator.id': id,
        } as ActionParams,
        step,
        includeData: true,
        includeImmutableData: false,
        dataSe: getAuthenticatorDataSeVal(AUTHENTICATOR_KEY.OV, option.value as string),
        iconName: option.value === 'totp' ? `oktaVerify_${index}` : `oktaVerifyPush_${index}`,
        iconDescr: option.value === 'totp'
          ? loc('factor.totpSoft.description', 'login')
          : loc('factor.push.description', 'login'),
      },
    };
    return authenticatorButton;
  });
};

const getAuthenticatorDescriptionParams = (
  option: IdxOption,
  authenticatorKey: string,
  isEnroll?: boolean,
): string[] | undefined => {
  const authenticatorDescrWithParams = [
    AUTHENTICATOR_KEY.ON_PREM,
    AUTHENTICATOR_KEY.IDP,
    AUTHENTICATOR_KEY.CUSTOM_APP,
    AUTHENTICATOR_KEY.SYMANTEC_VIP,
  ];
  if (!isEnroll || !authenticatorDescrWithParams.includes(authenticatorKey)) {
    return undefined;
  }

  switch (authenticatorKey) {
    case AUTHENTICATOR_KEY.ON_PREM: {
      const vendorName = option.relatesTo?.displayName
        || loc('oie.on_prem.authenticator.default.vendorName', 'login');
      return [vendorName];
    }
    case AUTHENTICATOR_KEY.IDP: {
      const idpName = option.relatesTo?.displayName || '';
      return [idpName];
    }
    case AUTHENTICATOR_KEY.CUSTOM_APP: {
      const customAppName = option.label || '';
      return [customAppName];
    }
    case AUTHENTICATOR_KEY.SYMANTEC_VIP: {
      const appName = option.relatesTo?.displayName || '';
      return [appName];
    }
    default:
      return undefined;
  }
};

const getAuthenticatorDescription = (
  option: IdxOption,
  authenticatorKey: string,
  isEnroll?: boolean,
): string | undefined => {
  if (!authenticatorKey) {
    return undefined;
  }
  const descrParams = getAuthenticatorDescriptionParams(
    option,
    authenticatorKey,
    isEnroll,
  );
  if (isEnroll) {
    return loc(AUTHENTICATOR_ENROLLMENT_DESCR_KEY_MAP[authenticatorKey], 'login', descrParams);
  }

  switch (authenticatorKey) {
    case AUTHENTICATOR_KEY.PHONE:
      return option.relatesTo?.profile?.phoneNumber as string || undefined;
    case AUTHENTICATOR_KEY.EMAIL:
      return option.relatesTo?.profile?.email as string || undefined;
    case AUTHENTICATOR_KEY.CUSTOM_APP:
      return option.relatesTo?.displayName as string || undefined;
    case AUTHENTICATOR_KEY.OV:
      return loc('oie.okta_verify.label', 'login');
    default:
      return undefined;
  }
};

const getNickname = (
  option: IdxOption,
  authenticatorKey: string,
  isEnroll?: boolean,
): string | undefined => {
  if (!authenticatorKey || isEnroll) {
    return undefined;
  }

  if (authenticatorKey === AUTHENTICATOR_KEY.PHONE) {
    // @ts-expect-error OKTA-661650 nickname missing from IdxAuthenticator
    return option.relatesTo?.nickname;
  }
  return undefined;
};

const getUsageDescription = (option: IdxOption): string | undefined => {
  // @ts-ignore IdxAuthenticator missing allowedFor property
  const { allowedFor } = option.relatesTo;
  switch (allowedFor) {
    case AUTHENTICATOR_ALLOWED_FOR_OPTIONS.ANY:
      return loc('oie.enroll.authenticator.usage.text.access.recovery', 'login');
    case AUTHENTICATOR_ALLOWED_FOR_OPTIONS.RECOVERY:
      return loc('oie.enroll.authenticator.usage.text.recovery', 'login');
    case AUTHENTICATOR_ALLOWED_FOR_OPTIONS.SSO:
      return loc('oie.enroll.authenticator.usage.text.access', 'login');
    default:
      return undefined;
  }
};

const formatAuthenticatorOptions = (
  options: IdxOption[],
  step: string,
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const authenticatorOptionSet = new Set<string>();
  return options
    .filter((option: IdxOption) => {
      if (isEnroll) {
        return true;
      }
      // If webauthn enrollments > 1 just show one entry with a generic namne (first)
      // so user doesnt have to select which one to pick. eg)
      // If there is yubikey5 and another unknown u2f key, user cannot identify that easily.
      // We need to do this at least  until users can give authenticator enrollments custom names.
      let isDup = false;
      const authenticatorKey = option.relatesTo?.key as string;
      if (authenticatorKey === AUTHENTICATOR_KEY.WEBAUTHN) {
        isDup = authenticatorOptionSet.has(authenticatorKey);
        authenticatorOptionSet.add(authenticatorKey);
      } else if (authenticatorKey === AUTHENTICATOR_KEY.CUSTOM_APP) {
        const id = getOptionValue(option.value as Input[], 'id')?.value as string;
        isDup = authenticatorOptionSet.has(id);
        authenticatorOptionSet.add(id);
      }
      return !isDup;
    })
    .map((option: IdxOption, index: number) => {
      const authenticatorKey = option.relatesTo?.key as string;
      const id = getOptionValue(option.value as Input[], 'id')?.value;
      const methodType = getOptionValue(option.value as Input[], 'methodType')?.value;
      const enrollmentId = getOptionValue(option.value as Input[], 'enrollmentId')?.value;
      const AUTHENTICATORS_WITH_METHOD_TYPE = [
        AUTHENTICATOR_KEY.ON_PREM,
        AUTHENTICATOR_KEY.OV,
        AUTHENTICATOR_KEY.RSA,
      ];
      const AUTHENTICATORS_WITH_NO_TRANSLATE_CLASS = [
        AUTHENTICATOR_KEY.PHONE,
        AUTHENTICATOR_KEY.CUSTOM_APP,
      ];
      const AUTHENTICATORS_WITH_LTR_DESCRIPTION = [
        AUTHENTICATOR_KEY.PHONE,
      ];
      const authenticator = option.relatesTo;

      return {
        type: 'AuthenticatorButton',
        label: option.label,
        id: `auth_btn_${authenticatorKey}_${enrollmentId || id}`,
        noTranslate: !isEnroll && AUTHENTICATORS_WITH_NO_TRANSLATE_CLASS.includes(authenticatorKey),
        dir: !isEnroll && AUTHENTICATORS_WITH_LTR_DESCRIPTION.includes(authenticatorKey) ? 'ltr' : undefined,
        options: {
          type: ButtonType.BUTTON,
          key: authenticatorKey,
          authenticator,
          ctaLabel: isEnroll
            ? loc('oie.enroll.authenticator.button.text', 'login')
            : loc('oie.verify.authenticator.button.text', 'login'),
          description: getAuthenticatorDescription(
            option,
            authenticatorKey,
            isEnroll,
          ),
          nickname: getNickname(option, authenticatorKey, isEnroll),
          usageDescription: isEnroll && getUsageDescription(option),
          // @ts-ignore logoUri missing from interface
          logoUri: authenticator.logoUri,
          actionParams: {
            'authenticator.id': id,
            'authenticator.methodType': AUTHENTICATORS_WITH_METHOD_TYPE.includes(authenticatorKey)
              ? methodType
              : undefined,
            'authenticator.enrollmentId': enrollmentId,
          },
          step,
          includeData: true,
          includeImmutableData: false,
          dataSe: getAuthenticatorDataSeVal(
            authenticatorKey,
            AUTHENTICATORS_WITH_METHOD_TYPE.includes(authenticatorKey) && typeof methodType === 'string'
              ? methodType
              : undefined,
          ),
          iconName: `${authenticatorKey}_${index}`,
        },
      } as AuthenticatorButtonElement;
    });
};

const getAuthenticatorButtonElements = (
  options: IdxOption[],
  step: string,
  isEnroll?: boolean,
): AuthenticatorButtonElement[] => {
  const formattedOptions = formatAuthenticatorOptions(options, step, isEnroll);

  // appending OV options back to its original spot
  const ovOptions = buildOktaVerifyOptions(options, step, isEnroll);
  if (ovOptions.length && options?.length) {
    const ovIndex = options.findIndex(({ relatesTo }) => relatesTo?.key === AUTHENTICATOR_KEY.OV);
    formattedOptions.splice(ovIndex, 1, ...ovOptions);
  }

  const ovRemediation = options.find((option) => option.relatesTo?.key === AUTHENTICATOR_KEY.OV);
  const methodType = (ovRemediation?.value as Input[])?.find(({ name }) => name === 'methodType');
  // Only re-order auth buttons when options contains OV with signed_nonce method type
  if (!methodType?.options?.find((option: IdxOption) => option.value === 'signed_nonce')) {
    return formattedOptions;
  }
  const deviceKnown = ovRemediation?.relatesTo?.deviceKnown;
  return reorderAuthenticatorButtons(formattedOptions, deviceKnown);
};

export const getAppAuthenticatorMethodButtonElements = (
  authenticator: Input,
  step: string,
  authKey = AUTHENTICATOR_KEY.OV,
  deviceKnown?: boolean,
): AuthenticatorButtonElement[] => {
  const id = (authenticator.value as Input[])?.find(({ name }) => name === 'id')?.value as string;
  const methodType = (authenticator.value as Input[])?.find(({ name }) => name === 'methodType');
  if (!methodType?.options?.length) {
    return [];
  }

  const authButtons = methodType.options.map((option, index) => ({
    type: 'AuthenticatorButton',
    label: option.label,
    id: `auth_btn_${authKey}_${option.value as string}`,
    noTranslate: authKey === AUTHENTICATOR_KEY.CUSTOM_APP,
    options: {
      type: ButtonType.BUTTON,
      key: authKey,
      ctaLabel: loc('oie.verify.authenticator.button.text', 'login'),
      actionParams: {
        'authenticator.id': id,
        'authenticator.methodType': (option.value as string),
      },
      description: getAuthenticatorDescription(
        option,
        authKey,
        false,
      ),
      dataSe: getAuthenticatorDataSeVal(
        authKey,
        option.value as string,
      ),
      iconName: `${authKey}_${index}`,
      step,
      includeData: true,
      includeImmutableData: false,
    },
  })) as AuthenticatorButtonElement[];

  return reorderAuthenticatorButtons(authButtons, deviceKnown);
};

export const isOnlyPushWithAutoChallenge = (
  inputs?: Input[],
): boolean => {
  const methodType = inputs?.find(({ name }) => name === 'methodType');
  const autoChallenge = inputs?.find(({ name }) => name === 'autoChallenge');

  return typeof autoChallenge !== 'undefined'
    && methodType?.options?.length === 1
    && methodType.options[0].value === 'push';
};

export const getAuthenticatorVerifyButtonElements = (
  authenticatorOptions: IdxOption[],
  step: string,
):AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
  step,
);

export const getAuthenticatorEnrollButtonElements = (
  authenticatorOptions: IdxOption[],
  step: string,
): AuthenticatorButtonElement[] => getAuthenticatorButtonElements(
  authenticatorOptions,
  step,
  true,
);
