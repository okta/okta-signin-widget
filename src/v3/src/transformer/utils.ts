/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FieldElement,
  FormBag,
  UISchemaElement,
  UISchemaLayoutType,
} from '../types';

export const createForm = (): FormBag => ({
  schema: {},
  uischema: {
    type: UISchemaLayoutType.VERTICAL,
    elements: [],
  },
  data: {},
  dataSchema: {
    fieldsToValidate: [] as string[],
  } as FormBag['dataSchema'],
});

export const removeUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement[] => (
  elements.filter((element) => (
    name !== (element as FieldElement).options?.inputMeta?.name
  ))
);

export const getUIElementWithName = (
  name: string,
  elements: UISchemaElement[],
): UISchemaElement | undefined => (
  elements.find((element) => (
    name === (element as FieldElement).options?.inputMeta?.name
  ))
);
