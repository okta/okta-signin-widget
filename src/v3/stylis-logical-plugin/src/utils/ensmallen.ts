import { compile, serialize, stringify } from 'stylis';

export const ensmallen = (css: string) => {
  return serialize(compile(css), stringify);
};
