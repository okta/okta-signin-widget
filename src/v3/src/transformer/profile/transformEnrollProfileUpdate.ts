/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  ButtonElement,
  ButtonType,
  DescriptionElement,
  FieldElement,
  IdxStepTransformer,
  LinkElement,
  TitleElement,
  UISchemaElement,
} from '../../types';
import { loc } from '../../util';
import { getUIElementWithName } from '../utils';

const SECOND_EMAIL_FIELD_NAME = 'userProfile.secondEmail';

export const transformEnrollProfileUpdate: IdxStepTransformer = ({ transaction, formBag }) => {
  const { availableSteps, nextStep } = transaction;
  const { uischema, data } = formBag;

  const secondEmailField = getUIElementWithName(
    SECOND_EMAIL_FIELD_NAME,
    uischema.elements as UISchemaElement[],
  );
  if (typeof secondEmailField !== 'undefined') {
    /**
     * As per requirement of this flow set secondEmail default to empty string,
     * if exists in remediation ideally server should have passed default string in remediation
     */
    data[SECOND_EMAIL_FIELD_NAME] = '';
  }

  const hasAllOptionalFields = uischema.elements
    .filter((ele) => ele.type === 'Field')
    .every((element) => (element as FieldElement).options.inputMeta.required === false);

  const titleElement: TitleElement = {
    type: 'Title',
    options: { content: loc('oie.profile.additional.title', 'login') },
  };

  const subtitleElement: DescriptionElement = {
    type: 'Description',
    contentType: 'subtitle',
    options: { content: loc('oie.form.field.optional.description', 'login') },
  };

  const submitBtnElement: ButtonElement = {
    type: 'Button',
    label: loc('enroll.choices.submit.finish', 'login'),
    options: {
      type: ButtonType.SUBMIT,
      step: nextStep!.name,
    },
  };

  uischema.elements.unshift(subtitleElement);
  uischema.elements.unshift(titleElement);
  uischema.elements.push(submitBtnElement);

  const skipStep = availableSteps?.find(({ name }) => name === 'skip' );
  if (skipStep && hasAllOptionalFields) {
    const { name: step } = skipStep;
    const skipLinkEle: LinkElement = {
      type: 'Link',
      options: {
        label: loc('oie.enroll.skip.profile', 'login'),
        step,
        isActionStep: false,
      },
    };
    uischema.elements.push(skipLinkEle);
  }

  return formBag;
};
