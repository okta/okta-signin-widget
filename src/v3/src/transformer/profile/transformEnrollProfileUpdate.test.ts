/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { getStubFormBag, getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import {
  ButtonElement,
  ButtonType,
  DescriptionElement,
  FieldElement,
  LinkElement,
  TitleElement,
  WidgetProps,
} from 'src/types';

import { transformEnrollProfileUpdate } from './transformEnrollProfileUpdate';

describe('Enroll Profile Update Transformer Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  const formBag = getStubFormBag();
  let widgetProps: WidgetProps;
  beforeEach(() => {
    formBag.uischema.elements = [
      { type: 'Field', options: { inputMeta: { name: 'userProfile.firstName', required: false } } } as FieldElement,
      { type: 'Field', options: { inputMeta: { name: 'userProfile.lastName', required: false } } } as FieldElement,
      { type: 'Field', options: { inputMeta: { name: 'userProfile.email', required: false } } } as FieldElement,
    ];
    transaction.nextStep = {
      name: '',
    };
    transaction.availableSteps = [{ name: 'skip', action: jest.fn() }];
    widgetProps = {};
  });

  it('should add title, button, skip link and input elements from remediation with all optional fields', () => {
    const updatedFormBag = transformEnrollProfileUpdate({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.data['userProfile.secondEmail']).toBeUndefined();
    expect(updatedFormBag.uischema.elements.length).toBe(7);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.profile.additional.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect((updatedFormBag.uischema.elements[2] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).label)
      .toBe('enroll.choices.submit.finish');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
    expect(updatedFormBag.uischema.elements[6].type).toBe('Link');
    expect((updatedFormBag.uischema.elements[6] as LinkElement).options.label)
      .toBe('oie.enroll.skip.profile');
    expect((updatedFormBag.uischema.elements[6] as LinkElement).options.isActionStep).toBe(false);
  });

  it('should add title, button and input elements from remediation with all optional fields when skip is missing from remediation', () => {
    transaction.availableSteps = [];

    const updatedFormBag = transformEnrollProfileUpdate({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.data['userProfile.secondEmail']).toBeUndefined();
    expect(updatedFormBag.uischema.elements.length).toBe(6);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.profile.additional.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect((updatedFormBag.uischema.elements[2] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).label)
      .toBe('enroll.choices.submit.finish');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[5] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
  });

  it('should add title button and input elements from remediation with at least one required field', () => {
    formBag.uischema.elements.push({
      type: 'Field',
      label: 'Another field',
      options: {
        inputMeta: { name: 'userProfile.anotherField', required: true },
      },
    } as FieldElement);

    const updatedFormBag = transformEnrollProfileUpdate({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.data['userProfile.secondEmail']).toBeUndefined();
    expect(updatedFormBag.uischema.elements.length).toBe(7);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.profile.additional.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect((updatedFormBag.uischema.elements[2] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[5] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.anotherField');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).label)
      .toBe('enroll.choices.submit.finish');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
  });

  it('should add title button and input elements from remediation with at least one required field and secondEmail field', () => {
    formBag.uischema.elements.push({
      type: 'Field',
      label: 'Secondary email',
      options: {
        inputMeta: { name: 'userProfile.secondEmail', required: true },
      },
    } as FieldElement);

    const updatedFormBag = transformEnrollProfileUpdate({ transaction, formBag, widgetProps });

    expect(updatedFormBag).toMatchSnapshot();
    expect(updatedFormBag.data['userProfile.secondEmail']).toBe('');
    expect(updatedFormBag.uischema.elements.length).toBe(7);
    expect(updatedFormBag.uischema.elements[0].type).toBe('Title');
    expect((updatedFormBag.uischema.elements[0] as TitleElement).options?.content)
      .toBe('oie.profile.additional.title');
    expect((updatedFormBag.uischema.elements[1] as DescriptionElement).options?.content)
      .toBe('oie.form.field.optional.description');
    expect((updatedFormBag.uischema.elements[2] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.firstName');
    expect((updatedFormBag.uischema.elements[3] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.lastName');
    expect((updatedFormBag.uischema.elements[4] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.email');
    expect((updatedFormBag.uischema.elements[5] as FieldElement).options?.inputMeta.name)
      .toBe('userProfile.secondEmail');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).label)
      .toBe('enroll.choices.submit.finish');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).type).toBe('Button');
    expect((updatedFormBag.uischema.elements[6] as ButtonElement).options?.type)
      .toBe(ButtonType.SUBMIT);
  });
});
