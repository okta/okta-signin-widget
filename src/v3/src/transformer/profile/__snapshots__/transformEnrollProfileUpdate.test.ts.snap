// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enroll Profile Update Transformer Tests should add title button and input elements from remediation with at least one required field 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.profile.additional.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "Another field",
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.anotherField",
            "required": true,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "enroll.choices.submit.finish",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Enroll Profile Update Transformer Tests should add title button and input elements from remediation with at least one required field and secondEmail field 1`] = `
Object {
  "data": Object {
    "userProfile.secondEmail": "",
  },
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.profile.additional.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "Secondary email",
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.secondEmail",
            "required": true,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "enroll.choices.submit.finish",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Enroll Profile Update Transformer Tests should add title, button and input elements from remediation with all optional fields when skip is missing from remediation 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.profile.additional.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "enroll.choices.submit.finish",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`Enroll Profile Update Transformer Tests should add title, button, skip link and input elements from remediation with all optional fields 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {
    "fieldsToExclude": [Function],
    "fieldsToTrim": Array [],
    "fieldsToValidate": Array [],
    "submit": Object {
      "step": "challenge-authenticator",
    },
  },
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "options": Object {
          "content": "oie.profile.additional.title",
        },
        "type": "Title",
      },
      Object {
        "contentType": "subtitle",
        "options": Object {
          "content": "oie.form.field.optional.description",
        },
        "type": "Description",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.firstName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.lastName",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "options": Object {
          "inputMeta": Object {
            "name": "userProfile.email",
            "required": false,
          },
        },
        "type": "Field",
      },
      Object {
        "label": "enroll.choices.submit.finish",
        "options": Object {
          "step": "",
          "type": "submit",
        },
        "type": "Button",
      },
      Object {
        "options": Object {
          "isActionStep": false,
          "label": "oie.enroll.skip.profile",
          "step": "skip",
        },
        "type": "Link",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
